### PR TITLE
Add support for MongoDB datatypes: Decimal128 and ObjectId

### LINKED ISSUE #6712: Decimal128 data type
See RPM-88 for details.

### PR DESCRIPTION
**Pull request by cmelchior**
Closes #6712 
Closes https://github.com/realm/realm-java/issues/6711

This PR adds support for datatype specific to MongoDB, specifically `ObjectId` and `Decimal128`

## TODO
- [ ] Measure the impact of adding a dependency on the BSON library. We probably want to make it optional for non-sync no matter what, but we should also check the impact on the sync variant
- [ ] Core/ObjectId release to depend on

### Annotation processor
- [x] Allow `Decimal128` as a type in `ClassMetaData`. Indexable: No, Primary key: No
- [x] Allow `ObjectId` as a type in `ClassMetaData`. Indexable: Yes. Primary Key: Yes
- [x] Annotation processor tests
- [x] Verify how `RealmResults.toJson()` works with ObjectId and Decimal128.
- [x] `Realm.create(Class, ObjectId)`
- [x] `copyToRealm`
- [x] `copyToRealmOrUpdate`
- [x] `insert`
- [x] `insertOrUpdate`
- [x] `createObjectFromJson(Class, InputStream)`
- [x] `createObjectFromJson(Class, JSONObject)`
- [x] `createObjectFromJson(Class, String)`
- [ ] `createOrUpdateObjectFromJson(Class, InputStream)`
- [ ] `createOrUpdateObjectFromJson(Class, JSONObject)`
- [ ] `createOrUpdateObjectFromJson(Class, String)`
- [ ] `createAllFromJson(Class, InputStream)`
- [ ] `createAllFromJson(Class, JSONArray)`
- [ ] `createAllFromJson(Class, String)`
- [ ] `createOrUpdateAllFromJson(Class, InputStream)`
- [ ] `createOrUpdateAllFromJson(Class, JSONArray)`
- [ ] `createOrUpdateAllFromJson(Class, String)`

### DynamicRealmObject
- [x] `getDecimal128(field)`
- [x] `setDecimal128(field)`
- [x] `getObjectId(field)`
- [x] `setObjectId(field)`

### RealmResults
- [x] `set(field, Object)`
- [x] `setDecimal128(field, Decimal128)`
- [x] `setObjectId(field, ObjectId)`

### Query-predicates
- [x] Aggregates are not support for `Decimal128` (limitation of Core)

### Decimal128
- [x] `equalTo`
- [x] `notEqualTo`
- [x] `lessThan`
- [x] `lessThanOrEqual`
- [x]  `greater`
- [x]  `greaterThan`

### ObjectId
- [x] `equalTo`
- [x] `notEqualTo`
- [x] `lessThan`
- [x] `lessThanOrEqual`
- [x]  `greater`
- [x]  `greaterThan`

### CODE REVIEW

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/Constants.kt [hunk 1]
```diff
@@ -70,34 +74,42 @@ object Constants {
...
-        JAVA_TO_REALM_TYPES["java.lang.Long"] = RealmFieldType.INTEGER
-        JAVA_TO_REALM_TYPES["java.lang.Float"] = RealmFieldType.FLOAT
-        JAVA_TO_REALM_TYPES["java.lang.Double"] = RealmFieldType.DOUBLE
-        JAVA_TO_REALM_TYPES["java.lang.Boolean"] = RealmFieldType.BOOLEAN
-        JAVA_TO_REALM_TYPES["java.lang.String"] = RealmFieldType.STRING
-        JAVA_TO_REALM_TYPES["java.util.Date"] = RealmFieldType.DATE
-        JAVA_TO_REALM_TYPES["byte[]"] = RealmFieldType.BINARY
-        // TODO: add support for char and Char
+        JAVA_TO_REALM_TYPES.apply {
+            this["byte"] = RealmFieldType.INTEGER
```
Comment (Zhuinden) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/Constants.kt]: You could do `JAVA_TO_REALM_TYPES = mapOf("byte" to RealmFieldType.INTEGER)` and so on to make it more Kotlin-y

File: realm/kotlin-extensions/build.gradle [hunk 2]
```diff
@@ -70,6 +70,7 @@ android {
 }
 
 dependencies {
+    compileOnly 'org.mongodb:bson:3.12.0'
```
Comment (nhachicha) [inline realm/kotlin-extensions/build.gradle]: why adding this, `realm-library` already includes `implementation 'org.mongodb:bson:3.12.0'` ?

File: realm/realm-library/src/main/java/io/realm/DynamicRealmObject.java [hunk 3]
```diff
@@ -558,6 +568,13 @@ private void setValue(String fieldName, Object value) {
             RealmList<?> list = (RealmList<?>) value;
             setList(fieldName, list);
         } else {
+            // Handle Decimal128 and ObjectId in a special way since they might not be on the
+            // classpath
```
Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/DynamicRealmObject.java]: I'm not sure about this comment, you added the Bson deps in the library gradle `implementation 'org.mongodb:bson:3.12.0'`

File: realm/realm-library/src/testUtils/java/io/realm/TestHelper.java [hunk 4]
```diff
@@ -1304,4 +1306,26 @@ public static void waitForNetworkThreadExecutorToFinish() {
...
+        char[] hex = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E' , 'F'};
+
+        StringBuilder randomId = new StringBuilder(24);
+        for (int j = 0; j < 24; j++) {
+            randomId.append(hex[(i + j) % 16]);
+        }
+        return randomId.toString();
+    }
+
+
```
Comment (cmelchior) [inline realm/realm-library/src/testUtils/java/io/realm/TestHelper.java:1330]:
```suggestion
```

File: realm/realm-library/src/testUtils/java/io/realm/TestHelper.java [hunk 5]
```diff
@@ -76,7 +78,7 @@
     public static final int STANDARD_WAIT_SECS = 200;
 
     private static final Charset UTF_8 = Charset.forName("UTF-8");
-    private static final Random RANDOM = new Random();
+    private static final SecureRandom RANDOM = new SecureRandom();
```
Comment (cmelchior) [inline realm/realm-library/src/testUtils/java/io/realm/TestHelper.java:81]: Is SecureRandom more random than Random?

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp [hunk 6]
```diff
@@ -1503,6 +1896,8 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeIsNotNull(JNIEnv*
                 case type_Float:
                 case type_Double:
                 case type_Timestamp:
+                case type_Decimal:
+                case type_ObjectId:
                     pQuery->not_equal(ColKey(column_idx), realm::null());
                     break;
                 default:
```
Comment (bmunkholm) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp:1903]: Unrelated to the PR, I just happened to notice: Why are we having "default:" here (and many other places) - it's better not to to ensure the compiler throws a warning when we are missing the handling of a new enum.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmJsonTypeHelper.kt [hunk 7]
```diff
@@ -318,6 +317,97 @@ object RealmJsonTypeHelper {
...
+    private class Decimal128TypeConverter : JsonToRealmFieldTypeConverter {
+        @Throws(IOException::class)
+        override fun emitTypeConversion(varName: String, accessor: String, fieldName: String, fieldType: QualifiedClassName, writer: JavaWriter) {
+            writer.apply {
+                beginControlFlow("if (json.has(\"%s\"))", fieldName)
+                beginControlFlow("if (json.isNull(\"%s\"))", fieldName)
+                emitStatement("%s.%s(null)", varName, accessor)
+                nextControlFlow("else")
+                emitStatement("Object decimal = json.get(\"%s\")", fieldName)
+                beginControlFlow("if (decimal instanceof org.bson.types.Decimal128)")
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmJsonTypeHelper.kt:329]: yes it's tested in `createUsingJsonStream_decimal128`

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmJsonTypeHelper.kt [hunk 8]
```diff
@@ -318,6 +317,97 @@ object RealmJsonTypeHelper {
...
+    private class ObjectIdTypeConverter() : JsonToRealmFieldTypeConverter {
+        @Throws(IOException::class)
+        override fun emitTypeConversion(varName: String, accessor: String, fieldName: String, fieldType: QualifiedClassName, writer: JavaWriter) {
+            writer.apply {
+                beginControlFlow("if (json.has(\"%s\"))", fieldName)
+                beginControlFlow("if (json.isNull(\"%s\"))", fieldName)
+                emitStatement("%s.%s(null)", varName, accessor)
+                nextControlFlow("else")
+                emitStatement("Object id = json.get(\"%s\")", fieldName)
+                beginControlFlow("if (id instanceof org.bson.types.ObjectId)")
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmJsonTypeHelper.kt:370]: it is tested in `createObjectFromJson_objectId`

File: realm/realm-library/src/androidTest/java/io/realm/RealmJsonTests.java [hunk 9]
```diff
@@ -247,6 +251,86 @@ public void createObjectFromJson_dateAsString() throws JSONException {
         assertEquals(new Date(1000), obj.getColumnDate());
     }
 
+    @Test
+    public void createObjectFromJson_decimal128() throws JSONException {
+        JSONObject json = new JSONObject();
+        json.put("columnDecimal128", new Decimal128(BigDecimal.TEN));
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-library/src/androidTest/java/io/realm/RealmJsonTests.java:257]: Well you could write your own parser which produces a `JSONObject/array` it could be useful to add directly the Bson types if you know that the underlying consumer system can consume (parse them)

File: realm/realm-library/src/androidTest/java/io/realm/RealmJsonTests.java [hunk 10]
```diff
@@ -247,6 +251,86 @@ public void createObjectFromJson_dateAsString() throws JSONException {
...
+
+        AllTypes obj = realm.where(AllTypes.class).findFirst();
+        assertEquals(Decimal128.parse("32361122672259149"), obj.getColumnDecimal128());
+    }
+
+    @Test
+    public void createObjectFromJson_objectId() throws JSONException {
+        JSONObject json = new JSONObject();
+        String idHex = TestHelper.generateObjectIdHexString(7);
+        json.put("columnObjectId", new ObjectId(idHex));
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-library/src/androidTest/java/io/realm/RealmJsonTests.java:310]: see previous response
>  you could write your own parser which produces a JSONObject/array it could be useful to add directly the Bson types if you know that the underlying consumer system can consume (parse them)

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.kt [hunk 11]
```diff
@@ -1731,28 +1781,38 @@ class RealmProxyClassGenerator(private val processingEnvironment: ProcessingEnvi
...
+                var jsonAccessorMethodSuffix = "Long"
+                var findFirstCast = ""
+                if (Utils.isString(metadata.primaryKey)) {
+                    pkType = "String"
+                    jsonAccessorMethodSuffix=  "String"
+                } else if (Utils.isObjectId(metadata.primaryKey)) {
+                    pkType = "ObjectId"
+                    findFirstCast = "(org.bson.types.ObjectId)"
+                    jsonAccessorMethodSuffix = ""
+                }
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.kt:1794]: The else case is actually the `Long` use case, so we shouldn't throw here

File: realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java [hunk 12]
```diff
@@ -652,6 +659,83 @@ public void equalTo() {
...
+        RealmResults<AllTypes> resultList = realm.where(AllTypes.class).sort(AllTypes.FIELD_OBJECT_ID, Sort.ASCENDING).findAll();
+        for (int i = 0; i < 10; i++) {
+            assertEquals(new ObjectId(TestHelper.generateObjectIdHexString(i)), resultList.get(i).getColumnObjectId());
+        }
+    }
+
+    @Test
+    public void notEqualTo_decimal128() {
+        populateTestRealm(realm, 10);
+        // FIXME
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java:683]: These are covered in `Decimal128Tests` and `ObjectIdTests` respectively

File: realm/realm-library/src/androidTest/kotlin/io/realm/Decimal128Tests.kt [hunk 13]
```diff
@@ -0,0 +1,437 @@
...
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.rules.TemporaryFolder
+import org.junit.runner.RunWith
+import java.math.BigDecimal
+
+open class Decimal128Required
+ : RealmObject() {
+    @field:PrimaryKey
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-library/src/androidTest/kotlin/io/realm/Decimal128Tests.kt:21]: It's not required just to limit the scope of the annotation

File: realm/realm-library/src/androidTest/kotlin/io/realm/Decimal128Tests.kt [hunk 14]
```diff
@@ -0,0 +1,437 @@
...
+        realm = Realm.getInstance(realmConfiguration)
+    }
+
+    @After
+    fun tearDown() {
+        realm.close()
+    }
+
+    @Test
+    fun copyToAndFromRealm() {
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-library/src/androidTest/kotlin/io/realm/Decimal128Tests.kt:89]: updated

File: realm/realm-library/src/androidTest/kotlin/io/realm/Decimal128Tests.kt [hunk 15]
```diff
@@ -0,0 +1,437 @@
...
+        realm.commitTransaction()
+
+        assertEquals(2, realm.where<Decimal128OptionalRealmList>().findFirst()?.decimals?.size)
+    }
+
+    @Test
+    fun linkQueryNotSupported() {
+        try {
+            realm.where<Decimal128RequiredRealmList>().greaterThan("decimals", Decimal128(BigDecimal.ZERO)).findAll()
+            fail("It should not be possible to perform link query on Decimal128")
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-library/src/androidTest/kotlin/io/realm/Decimal128Tests.kt:198]: We don't support querying RealmList  yet (that's the case also for Date for example)

File: realm/realm-library/src/androidTest/kotlin/io/realm/ObjectIdTests.kt [hunk 16]
```diff
@@ -0,0 +1,352 @@
...
+        realm.commitTransaction()
+
+        assertEquals(2, realm.where<ObjectIdOptionalRealmList>().findFirst()?.ids?.size)
+    }
+
+    @Test
+    fun linkQueryNotSupported() {
+        try {
+            realm.where<ObjectIdRequiredRealmList>().greaterThan("ids", ObjectId(generateObjectIdHexString(0))).findAll()
+            fail("It should not be possible to perform link query on ObjectId")
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-library/src/androidTest/kotlin/io/realm/ObjectIdTests.kt:197]: This is the current case for Date for instance and other native types are not supported or WIP. I'm not sure where this is tracked in Core/OS

File: realm/realm-library/src/main/cpp/io_realm_internal_OsObject.cpp [hunk 17]
```diff
@@ -214,6 +214,37 @@ static inline Obj do_create_row_with_primary_key(JNIEnv* env, jlong shared_realm
...
+    JStringAccessor str_accessor(env, pk_value); // throws
+    if (!pk_value && !COL_NULLABLE(env, table, pk_column_key)) {
+        return Obj();
+    }
+
+
+    if (pk_value) {
+        auto objectId = ObjectId(StringData(str_accessor).data());
+        if (bool(table->find_first_object_id(col_key, objectId))) {
+            THROW_JAVA_EXCEPTION(env, PK_CONSTRAINT_EXCEPTION_CLASS,
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-library/src/main/cpp/io_realm_internal_OsObject.cpp:233]: It's unreachable code beyond this since it throws a `std::runtime_error` exception

File: realm/realm-library/src/main/cpp/io_realm_internal_OsObject.cpp [hunk 18]
```diff
@@ -214,6 +214,37 @@ static inline Obj do_create_row_with_primary_key(JNIEnv* env, jlong shared_realm
...
+                                 format(PK_EXCEPTION_MSG_FORMAT, str_accessor.operator std::string()));
+        }
+
+        return table->create_object_with_primary_key(objectId);
+    }
+    else {
+        if (bool(table->find_first_null(col_key))) {
+            THROW_JAVA_EXCEPTION(env, PK_CONSTRAINT_EXCEPTION_CLASS, format(PK_EXCEPTION_MSG_FORMAT, "'null'"));
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-library/src/main/cpp/io_realm_internal_OsObject.cpp:241]: It's unreachable code beyond this since it throws a `std::runtime_error` exception

File: realm/realm-library/src/main/cpp/io_realm_internal_Table.cpp [hunk 19]
```diff
@@ -540,6 +582,50 @@ JNIEXPORT void JNICALL Java_io_realm_internal_Table_nativeSetByteArray(JNIEnv* e
...
+                                                                    jlong columnKey, jlong rowKey, jstring j_value,
+                                                                    jboolean isDefault)
+{
+    TableRef table = TBL_REF(nativeTableRefPtr);
+    if (!TYPE_VALID(env, table, columnKey, type_ObjectId)) {
+        return;
+    }
+    try {
+        JStringAccessor value(env, j_value);
+        table->get_object(ObjKey(rowKey)).set(ColKey(columnKey), ObjectId(StringData(value).data()), B(isDefault));
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-library/src/main/cpp/io_realm_internal_Table.cpp:623]: no, it predates my era :)

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp [hunk 20]
```diff
@@ -1320,6 +1675,36 @@ JNIEXPORT jdouble JNICALL Java_io_realm_internal_TableQuery_nativeAverageDouble(
...
+                                                                                jlong columnKey)
+{
+    Query* pQuery = Q(nativeQueryPtr);
+    ConstTableRef pTable = pQuery->get_table();
+    if (!TYPE_VALID(env, pTable, columnKey, type_Decimal)) {
+        return nullptr;
+    }
+    try {
+        Decimal128 decimal128 = pQuery->average_decimal128(ColKey(columnKey));
+        if (!decimal128.is_null()) {
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp:1689]: replaced with `RETURN_DECIMAL128_AS_JLONG_ARRAY__OR_NULL` macro

File: realm/realm-library/src/main/java/io/realm/RealmFieldType.java [hunk 21]
```diff
@@ -140,25 +145,21 @@ public boolean isValid(Object obj) {
                 return (obj instanceof Float);
             case CORE_TYPE_VALUE_DOUBLE:
                 return (obj instanceof Double);
+            case CORE_TYPE_VALUE_DECIMAL128:
+                return (obj instanceof Decimal128);
             case CORE_TYPE_VALUE_OBJECT:
-                return false;
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-library/src/main/java/io/realm/RealmFieldType.java:144]: nice catch!

**Comment by bmunkholm** [comment]
@nhachicha Cab you tick off what's done in the top checklist?

File: realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (cmelchior) [reply realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java:683]: True, so they should just be removed from here?

File: realm/realm-library/src/main/cpp/io_realm_internal_Table.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (cmelchior) [reply realm/realm-library/src/main/cpp/io_realm_internal_Table.cpp:623]: Huh, never seen it before. Oh well 😄

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 22]
```diff
@@ -1645,10 +1866,24 @@ public double average(String fieldName) {
...
+    /**
+     * Returns the average of a given field.
+     * Does not support dotted field notation.
+     *
+     * @param fieldName the field to calculate average on. Only Decimal128 fields is supported. For other number types consider using {@link #average(String)}.
+     * @return the average for the given field amongst objects in query results. This will be of type Decimal128. If no objects exist or they all have {@code null}
+     * as the value for the given field {@code 0} will be returned. When computing the average, objects with {@code null} values are ignored.
+     * @throws java.lang.IllegalArgumentException if the field is not a Decimal128 type.
+     */
+    public @Nullable Decimal128 averageDecimal128(String fieldName) {
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-library/src/main/java/io/realm/RealmQuery.java:1881]: Good catch, this is still missing from Core

File: realm/realm-library/src/testUtils/java/io/realm/TestHelper.java (hunk 5 above)
Comment (nhachicha) [reply realm/realm-library/src/testUtils/java/io/realm/TestHelper.java:81]: According to the documentation it's more secure when used in the context of cryptography, which is the case with `encryptionKey` parameter of the RealmConfiguration. This is more to set the example for developers

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (cmelchior) [reply realm/realm-library/src/main/java/io/realm/RealmQuery.java:1881]: Ok. No reason to hold this PR waiting for that. We can always add later.

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp (hunk 6 above)
Comment (nhachicha) [reply realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp:1903]: Historically this (i.e `REALM_UNREACHABLE`) has been used by Core, it could be interesting in some use case to use an extensive switch statement to ensure we're covering all branches, which in this example will be a fall through for `'type_OldTable', 'type_OldMixed', and 'type_OldDateTime`

File: realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java:683]: yes

File: realm/realm-library/src/main/java/io/realm/RealmResults.java [hunk 23]
```diff
@@ -516,8 +547,18 @@ private Row checkRealmObjectConstraints(String fieldName, @Nullable RealmModel v
                 checkTypeOfListElements(list, Double.class);
                 osResults.setDoubleList(fieldName, (RealmList<Double>) list);
                 break;
-            default:
+            default: {
+                // Handle Decimal128 and ObjectId in a special way since they might not be on the
+                // classpath
+                if (columnType == RealmFieldType.DECIMAL128_LIST) {
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-library/src/main/java/io/realm/RealmResults.java]: As discussed, let's add the bson dependency for this PR and optimise later a mechanism to make it optional. I think the method count and size will not affect apps not using these types (because of R8 tree shaking)

File: realm/realm-library/src/main/java/io/realm/RealmList.java [hunk 24]
```diff
@@ -1328,6 +1331,14 @@ private static boolean isClassForRealmModel(Class<?> clazz) {
             //noinspection unchecked
             return (ManagedListOperator<E>) new DateListOperator(realm, osList, (Class<Date>) clazz);
         }
+        if (clazz == Decimal128.class) {
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-library/src/main/java/io/realm/RealmList.java:1334]: As discussed, let's add the bson dependency for this PR and optimise later a mechanism to make it optional. I think the method count and size will not affect apps not using these types (because of R8 tree shaking)

File: realm/realm-library/src/main/java/io/realm/DynamicRealmObject.java [hunk 25]
```diff
@@ -406,6 +451,10 @@ public DynamicRealmObject getObject(String fieldName) {
             return RealmFieldType.FLOAT_LIST;
         } else if (primitiveType.equals(Double.class)) {
             return RealmFieldType.DOUBLE_LIST;
+        } else if (primitiveType.equals(Decimal128.class)) {
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-library/src/main/java/io/realm/DynamicRealmObject.java:454]: As discussed, let's add the bson dependency for this PR and optimise later a mechanism to make it optional. I think the method count and size will not affect apps not using these types (because of R8 tree shaking)

File: realm/realm-library/src/main/java/io/realm/DynamicRealmObject.java [hunk 26]
```diff
@@ -557,6 +616,10 @@ private void setValue(String fieldName, Object value) {
         } else if (valueClass == RealmList.class) {
             RealmList<?> list = (RealmList<?>) value;
             setList(fieldName, list);
+        } else if (valueClass == Decimal128.class) {
```
[the remark this answers is not in the document]
Comment (nhachicha) [reply realm/realm-library/src/main/java/io/realm/DynamicRealmObject.java:619]: As discussed, let's add the bson dependency for this PR and optimise later a mechanism to make it optional. I think the method count and size will not affect apps not using these types (because of R8 tree shaking)

**Comment by nhachicha** [comment]
@cmelchior  Here are some stats about including the new Bson type as a dependency for all flavour:

**TL;DR**
There's no significant increase when using R8 to shrink and tree-shake the APK.

The comparison was done against `next-major` branch (https://github.com/realm/realm-java/commit/15424125c8c7ce0c092b7c87ee49c9d50ece2723) using the `IntroExample`, This branch was modified locally to use the same Android build tool and Gradle plugin version, and target the same minSDK 
```Diff
diff --git a/dependencies.list b/dependencies.list
index f80a11315..fc1f46e85 100644
--- a/dependencies.list
+++ b/dependencies.list
@@ -8,12 +8,12 @@ REALM_SYNC_SHA256=bccf1fb89e32950a78dca4e93074f6479e0d016320897ab8bd1135d8568a18
 REALM_OBJECT_SERVER_VERSION=3.28.2
 
 # Common Android settings across projects
-GRADLE_BUILD_TOOLS=3.3.2
-ANDROID_BUILD_TOOLS=28.0.3
+GRADLE_BUILD_TOOLS=3.6.1
+ANDROID_BUILD_TOOLS=29.0.2
 
 # Common classpath dependencies
 # Gradle 5 is not supported yet: https://issuetracker.google.com/issues/126433059
-gradleVersion=4.10.1
+gradleVersion=5.6.4
 ndkVersion=21.0.6113669
 BUILD_INFO_EXTRACTOR_GRADLE=4.7.5
 GRADLE_BINTRAY_PLUGIN=1.8.4
diff --git a/examples/build.gradle b/examples/build.gradle
index 043627766..49d6207ca 100644
--- a/examples/build.gradle
+++ b/examples/build.gradle
@@ -1,7 +1,7 @@
 def projectDependencies = new Properties()
 projectDependencies.load(new FileInputStream("${rootDir}/../dependencies.list"))
 project.ext.sdkVersion = 27
-project.ext.minSdkVersion = 16
+project.ext.minSdkVersion = 21
 project.ext.buildTools = projectDependencies.get("ANDROID_BUILD_TOOLS")
 
 // Don't cache SNAPSHOT (changing) dependencies.
diff --git a/examples/gradle/wrapper/gradle-wrapper.properties b/examples/gradle/wrapper/gradle-wrapper.properties
index 4e974715f..0ebb3108e 100644
--- a/examples/gradle/wrapper/gradle-wrapper.properties
+++ b/examples/gradle/wrapper/gradle-wrapper.properties
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.10.1-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.6.4-all.zip
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
diff --git a/examples/introExample/build.gradle b/examples/introExample/build.gradle
index 84188bec3..a5f986059 100644
--- a/examples/introExample/build.gradle
+++ b/examples/introExample/build.gradle
@@ -19,11 +19,27 @@ android {
 
     buildTypes {
         release {
+//            minifyEnabled true
+//            signingConfig signingConfigs.debug
+
+
+            // Enables code shrinking, obfuscation, and optimization for only
+            // your project's release build type.
             minifyEnabled true
-            signingConfig signingConfigs.debug
+
+            // Enables resource shrinking, which is performed by the
+            // Android Gradle plugin.
+            shrinkResources true
+
+            // Includes the default ProGuard rules files that are packaged with
+            // the Android Gradle plugin. To learn more, go to the section about
+            // R8 configuration files.
+            proguardFiles getDefaultProguardFile(
+                    'proguard-android-optimize.txt'),
+                    'proguard-rules.pro'
         }
         debug {
-            minifyEnabled true
+            minifyEnabled false
         }
     }
 
diff --git a/realm/realm-library/build.gradle b/realm/realm-library/build.gradle
index bc289641a..a3911c68c 100644
--- a/realm/realm-library/build.gradle
+++ b/realm/realm-library/build.gradle
@@ -58,7 +58,7 @@ android {
                     abiFilters(*project.getProperty('buildTargetABIs').trim().split('\\s*,\\s*'))
                 } else {
                     // "armeabi" and "mips" are no longer supported by the NDK
-                    abiFilters 'x86', 'x86_64', 'armeabi-v7a', 'arm64-v8a'
+                    abiFilters 'x86'
                 }
             }
         }

```

## Debug APKs

### NextMajor

IntroDebugNextMajor

### New types _(ObjectId, Decimal128)_

IntroDebugNexTypes

IntroDebugNewTypes


## Release APKs _(with R8)_

### NextMajor

IntroReleaseNextMajor

### New types (ObjectId, Decimal128)

IntroReleaseNewTypeUsage

## Comparison: old (NextMajor) vs new (NewTypes)

### Debug APK

DebugSizeCompareOldNextMajorVsNewTypes

### Release APK

ReleaseSizeCompareOldNextMajorVsNewTypes

## Comparison (New Type): Not using the types vs using _ObjectId and Decimal128_

### Release APK

IntroReleaseUsageNoUsageCompraison

**Comment by nhachicha** [comment]
Merging this while waiting for Core to add `sum` operator for `Decimal128`.
